### PR TITLE
Add coles template

### PR DESCRIPTION
**Pull request by mariaschuld**
**Context:**

Adding the ``SimplifiedTwoDesign`` template, which is an approximate version of a 2-design used in the Cerezo....Coles (2020) paper investigating barren plateau phenomena.

simplified_two_design

**Description of the Change:**

* add template file
* import template in module
* update ``introduction/templates.rst``
* add template tests
* add integration tests
* add init function for parameters
* add init tests

**Benefits:**

Growing the template library.

### CODE REVIEW

File: tests/test_init.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 """
-Unit tests for the :mod:`pennylane.templates.parameters` module.
+Unit tests for the :mod:`pennylane.init` module.
```
Comment (mariaschuld) [inline tests/test_init.py]: Found this error in the docstring...

File: tests/test_init.py [hunk 2]
```diff
@@ -129,26 +139,32 @@ def test_diff_output_for_diff_seed(self, init, sgntr, seed, tol):
         p1 = init(**s)
         s = {**s, 'seed': seed + 1}
         p2 = init(**s)
-        assert not np.allclose(p1, p2, atol=tol)
+
+        if p1.shape != (0,):
+            assert not np.allclose(p1, p2, atol=tol)
```
Comment (mariaschuld) [inline tests/test_init.py]: I realised that zero is an edge case here, so added this if statement.

File: tests/test_init.py [hunk 3]
```diff
@@ -129,26 +139,32 @@ def test_diff_output_for_diff_seed(self, init, sgntr, seed, tol):
...
     def test_interval(self, init, sgntr, seed, tol):
         """Test that sampled parameters lie in correct interval."""
-        s = {**sgntr, 'seed': seed}
 
-        # Case A: Uniformly distributed parameters
-        if 'low' in s.keys() and 'high' in s.keys():
-            s['low'] = 1
-            s['high'] = 1
+        # exclude case of empty parameter list
+        if not init(**sgntr).shape == (0, ):
```
Comment (mariaschuld) [inline tests/test_init.py]: Same here, avoiding the zero weights edge case.

File: pennylane/templates/layers/simplified_two_design.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    with respect to the Haar measure up to the first 2 moments.
+
+    The template is not a strict 2-design, since
+    it does not consist of universal 2-qubit gates as building blocks, but has been shown in
+    `Cerezo et al. (2020) <https://arxiv.org/abs/2001.00550>`_ to exhibit important properties to study "barren plateaus"
+    in quantum optimization landscapes.
+
+    The template starts with an initial layer of single qubit Pauli-Y rotations, before the main
+    :math:`L` layers are applied. The basic building block of the main layers are controlled-Z entanglers
+    followed by a pair of Pauli-Y gates.
```
Comment (antalszava) [inline pennylane/templates/layers/simplified_two_design.py]:
```suggestion
    followed by a Pauli-Y rotation gate on each wire.
```

File: pennylane/templates/layers/simplified_two_design.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    Each layer consists of an "even" part whose entanglers start with the first qubit,
+    and an "odd" part that starts with the second qubit.
+
+    This is an example of two layers, including the initial block:
+
+    .. figure:: ../../_static/templates/layers/simplified_two_design.png
+        :align: center
+        :width: 40%
+        :target: javascript:void(0);
```
Comment (antalszava) [inline pennylane/templates/layers/simplified_two_design.py]: Looks really nice! :1st_place_medal:

File: pennylane/templates/layers/simplified_two_design.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        initial_layer (array[float]): array of weights for the initial rotation block, shape ``(M,)``
+        weights (array[float]): array of rotation angles for the layers, shape ``(L, d, 2)``
+        wires (Sequence[int] or int): qubit indices that the template acts on
+
+    Raises:
+        ValueError: if inputs do not have the correct format
+
+    .. UsageDetails::
+
+        The template is used inside a qnode:
```
Comment (antalszava) [inline pennylane/templates/layers/simplified_two_design.py]: Linking
```suggestion
        The template with two layers is used inside a :class:`~.BaseQNode`:
```

File: pennylane/init.py [hunk 7]
```diff
@@ -835,3 +835,117 @@ def interferometer_varphi_normal(n_wires, mean=0, std=0.1, seed=None):
 
     varphi = np.random.normal(loc=mean, scale=std, size=(n_wires,))
     return varphi
+
+
+def simplified_two_design_initial_layer_uniform(n_wires, low=0, high=2 * pi, seed=None):
```
Comment (antalszava) [inline pennylane/init.py]: Here the name incorporates 3 things:
* name of the template
* the layer to get the parameters for
* distribution used when generating the parameters

Since some share common prefixes (all four, in this case, starts with `simplified_two_design`), it could be good to somehow change the structure here.

Possible ways to resolve this:
* Renaming using abbreviations
* Creating a general function that acts as a "factory" for template parameter initializer functions. E.g.: `get_initializer(template, block=None, distribution='uniform')` could be a signature. Not sure if this would overcomlicate things, but as the number of templates increases, it could be a good idea.

```python
initializer = get_initializer('simplified_two_design', block='initial', distribution='uniform')
params = initializer(n_wires=3)
```

File: pennylane/init.py [hunk 8]
```diff
@@ -835,3 +835,117 @@ def interferometer_varphi_normal(n_wires, mean=0, std=0.1, seed=None):
...
+    Returns:
+        array: parameter array
+    """
+    if seed is not None:
+        np.random.seed(seed)
+
+    n_unitaries_per_layer = n_wires // 2 + (n_wires - 1) // 2
+
+    if n_unitaries_per_layer == 0:
+        params = np.array([])
```
Comment (antalszava) [inline pennylane/init.py]: I'm assuming this is for convenience in the `n_wires=1` case, right?

File: tests/test_init.py [hunk 9]
```diff
@@ -43,6 +43,7 @@
 # Functions and their signatures
 
 # Functions returning a single parameter array
+# function name, kwargs and target shape
```
Comment (antalszava) [inline tests/test_init.py]: Really helpful comment :)

File: tests/test_init.py [hunk 10]
```diff
@@ -77,6 +78,16 @@
                       (qml.init.qaoa_embedding_uniform, {'n_layers': 2, 'n_wires': 2, 'low': 0, 'high': 1}, (2, 3)),
                       (qml.init.qaoa_embedding_normal, {'n_layers': 2, 'n_wires': 1, 'mean': 0, 'std': 1}, (2, 1)),
                       (qml.init.qaoa_embedding_normal, {'n_layers': 2, 'n_wires': 2, 'mean': 0, 'std': 1}, (2, 3)),
+                      (qml.init.simplified_two_design_initial_layer_uniform, {'n_wires': 1, 'low': 0, 'high': 1}, (1,)),
+                      (qml.init.simplified_two_design_initial_layer_uniform, {'n_wires': 3, 'low': 0, 'high': 1}, (3,)),
+                      (qml.init.simplified_two_design_initial_layer_normal, {'n_wires': 1, 'mean': 0, 'std': 1}, (1,)),
+                      (qml.init.simplified_two_design_initial_layer_normal, {'n_wires': 3, 'mean': 0, 'std': 1}, (3,)),
+                      (qml.init.simplified_two_design_weights_uniform, {'n_layers': 2, 'n_wires': 1, 'low': 0, 'high': 1}, (0,)),
```
Comment (antalszava) [inline tests/test_init.py]: Nice test case for this edge case!

File: tests/test_init.py (hunk 2 above)
Comment (antalszava) [reply tests/test_init.py]: Is this going to be an edge case for every template? If so, it could be organized into a separate test function. Not sure if it would make this file too big though. Anyways, porting the rest of the templates, in this case, would be out of scope and could be done in a subsequent PR.

File: tests/test_init.py [hunk 11]
```diff
@@ -129,26 +139,32 @@ def test_diff_output_for_diff_seed(self, init, sgntr, seed, tol):
...
-        # Case A: Uniformly distributed parameters
-        if 'low' in s.keys() and 'high' in s.keys():
-            s['low'] = 1
-            s['high'] = 1
+        # exclude case of empty parameter list
+        if not init(**sgntr).shape == (0, ):
+
+            s = {**sgntr, 'seed': seed}
+
+            # Case A: Uniformly distributed parameters
```
Comment (antalszava) [inline tests/test_init.py]: It could be nice to have the uniform and the normal cases separately. In such a case then one could see which fails exactly. I guess in tests it's fine to not worry about code duplication as long as it makes the test cases more explicit upon failing.

File: tests/test_templates.py [hunk 12]
```diff
@@ -281,6 +289,7 @@ def test_integration_qubit_diffable(self, template1, diffable1, nondiffable1,
                                         template2, diffable2, nondiffable2,
                                         interface, to_var):
         """Tests integration of qubit templates passing differentiable arguments as positional arguments to qnode."""
+        #TODO: rewrite test to avoid quadratic growth of test cases with the number of templates
```
Comment (antalszava) [inline tests/test_templates.py]: This could be the topic of a corresponding issue as well :)

File: tests/test_templates_layers.py [hunk 13]
```diff
@@ -462,3 +465,69 @@ def test_random_layer_weights(self, n_subsystems, tol):
...
+class TestSimplifiedTwoDesign:
+    """Tests for the SimplifiedTwoDesign method from the pennylane.templates.layers module."""
+
+    @pytest.mark.parametrize("n_wires, n_layers, shape_weights", [(1, 2, (0,)),
+                                                                  (2, 2, (2, 1, 2)),
+                                                                  (3, 2, (2, 2, 2)),
+                                                                  (4, 2, (2, 3, 2))])
+    def test_circuit_queue(self, n_wires, n_layers, shape_weights):
+        """Tests the gate types and parameter values in the circuit."""
+        np.random.seed(12)
```
Comment (antalszava) [inline tests/test_templates_layers.py]: Curious about this one :smile:

File: tests/test_templates_layers.py [hunk 14]
```diff
@@ -462,3 +465,69 @@ def test_random_layer_weights(self, n_subsystems, tol):
...
+
+
+class TestSimplifiedTwoDesign:
+    """Tests for the SimplifiedTwoDesign method from the pennylane.templates.layers module."""
+
+    @pytest.mark.parametrize("n_wires, n_layers, shape_weights", [(1, 2, (0,)),
+                                                                  (2, 2, (2, 1, 2)),
+                                                                  (3, 2, (2, 2, 2)),
+                                                                  (4, 2, (2, 3, 2))])
+    def test_circuit_queue(self, n_wires, n_layers, shape_weights):
```
Comment (antalszava) [inline tests/test_templates_layers.py]: Separated this into two tests: `test_circuit_queue_types` and `test_circuit_queue_parameter_values`, could make errors be more explicit.

File: tests/test_templates_layers.py [hunk 15]
```diff
@@ -462,3 +465,69 @@ def test_random_layer_weights(self, n_subsystems, tol):
...
+        # test the device parameters
+        for l in range(n_layers):
+            # only select the rotation gates
+            ops = [gate for gate in rec.queue if isinstance(gate, qml.RY)]
+
+            # check each initial_layer gate parameters
+            for n in range(n_wires):
+                res_param = ops[n].parameters[0]
+                exp_param = initial_layer[n]
+                assert res_param == exp_param
```
Comment (antalszava) [inline tests/test_templates_layers.py]: Nice :)

File: pennylane/templates/layers/simplified_two_design.py [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+         wires (list): two wire indices that unitary acts on
+    """
+
+    CZ(wires=wires)
+    RY(par1, wires=wires[0])
+    RY(par2, wires=wires[1])
+
+
+@template
+def SimplifiedTwoDesign(initial_layer, weights, wires):
```
Comment (antalszava) [inline pennylane/templates/layers/simplified_two_design.py]: The aim would be to make it apparent that here the parameters are accepted instead of the operations themselves (e.g. might be interpreted in a way that one has to give some sort of a generic layer structure).
```suggestion
def SimplifiedTwoDesign(initial_layer_weights, weights, wires):
```

File: pennylane/init.py [hunk 17]
```diff
@@ -835,3 +835,117 @@ def interferometer_varphi_normal(n_wires, mean=0, std=0.1, seed=None):
...
+        high (float): maximum value of uniform distribution
+        seed (int): seed used in sampling the parameters, makes function call deterministic
+
+    Returns:
+        array: parameter array
+    """
+    if seed is not None:
+        np.random.seed(seed)
+
+    n_unitaries_per_layer = n_wires // 2 + (n_wires - 1) // 2
```
Comment (thisac) [inline pennylane/init.py]: Isn't this just equal to `n_wires - 1`?

File: pennylane/init.py [hunk 18]
```diff
@@ -835,3 +835,117 @@ def interferometer_varphi_normal(n_wires, mean=0, std=0.1, seed=None):
...
+        std (float): standard deviation of parameters
+        seed (int): seed used in sampling the parameters, makes function call deterministic
+
+    Returns:
+        array: parameter array
+    """
+    if seed is not None:
+        np.random.seed(seed)
+
+    n_unitaries_per_layer = n_wires // 2 + (n_wires - 1) // 2
```
Comment (thisac) [inline pennylane/init.py]: Same as above: equal to `n_wires - 1`?

File: pennylane/templates/layers/simplified_two_design.py [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+        initial_layer,
+        expected_shape_initial,
+        msg="'initial_layer' must be of shape {}; got {}"
+        "".format(expected_shape_initial, _get_shape(initial_layer)),
+    )
+
+    if len(wires) in [0, 1]:
+        expected_shape_weights = (0,)
+    else:
+        expected_shape_weights = (repeat, len(wires) // 2 + (len(wires) - 1) // 2, 2)
```
Comment (thisac) [inline pennylane/templates/layers/simplified_two_design.py]: Same as above: `len(wires) // 2 + (len(wires) - 1) // 2` is equal to `len(wires) - 1`.

File: tests/test_init.py (hunk 2 above)
Comment (mariaschuld) [reply tests/test_init.py]: Good point, I'll mark it for another PR, this would make things very messy.

File: pennylane/templates/layers/simplified_two_design.py (hunk 4 above)
Comment (mariaschuld) [reply pennylane/templates/layers/simplified_two_design.py]:
```suggestion
    followed by a pair of Pauli-Y gates (one for each wire).
```

File: pennylane/templates/layers/simplified_two_design.py (hunk 6 above)
Comment (mariaschuld) [reply pennylane/templates/layers/simplified_two_design.py]: Could we do this one:

```suggestion
        The template - here shown for two layers - is used inside a :class:`~.BaseQNode`::
```

File: pennylane/init.py (hunk 7 above)
Comment (mariaschuld) [reply pennylane/init.py]: I totally agree! Initially, all init functions worked that way and had a flag "normal" or "uniform", but this became complicated to use (since one has to explain which input arguments are active for which flag), and we decided in favour of this structure.

An overhaul of the initialization functions is on our roadmap. The reason why it is a larger discussion is that it is so closely related to the templates data structure, which we may want to change into a class in the long run...

I suggest we leave it as is, and tackle it in a larger refactor after proper consultation.

File: pennylane/init.py (hunk 8 above)
Comment (mariaschuld) [reply pennylane/init.py]: Yes exactly!

File: tests/test_init.py (hunk 11 above)
Comment (mariaschuld) [reply tests/test_init.py]: Good point for future init test refactor!

File: pennylane/templates/layers/simplified_two_design.py (hunk 6 above)
Comment (josh146) [reply pennylane/templates/layers/simplified_two_design.py]: As per the other comment (I think on #555), we should link to the `qml.QNode` constructor, or `qml.qnode` decorator, here.

File: tests/test_templates.py (hunk 12 above)
Comment (mariaschuld) [reply tests/test_templates.py]: Added this as an issue.

File: tests/test_templates_layers.py (hunk 13 above)
Comment (mariaschuld) [reply tests/test_templates_layers.py]: Ah, another one of those copy and paste mistakes. Good find.

Comment (mariaschuld) [reply tests/test_templates_layers.py]: No wait, this was to initialise the random weights consistently.

File: tests/test_templates_layers.py (hunk 14 above)
Comment (mariaschuld) [reply tests/test_templates_layers.py]: Good point, I also sneaked in the split for the BasicEntangler tests.

File: pennylane/templates/layers/simplified_two_design.py (hunk 16 above)
Comment (mariaschuld) [reply pennylane/templates/layers/simplified_two_design.py]: Just that this name is so long...But I just can't think of a better name, so I changed it to your one...

File: pennylane/init.py (hunk 17 above)
Comment (mariaschuld) [reply pennylane/init.py]: Really? Oh wow, I had no idea, I am intrigued. That makes the code a lot easier!

File: pennylane/templates/layers/simplified_two_design.py (hunk 6 above)
Comment (mariaschuld) [reply pennylane/templates/layers/simplified_two_design.py]: Then:

```suggestion
        template - here shown for two layers - is used inside a :class:`~.QNode`:
```

File: .github/CHANGELOG.md [hunk 20]
```diff
@@ -5,7 +5,20 @@
...
+    :target: javascript:void(0);
+
+* Added the ``BasicEntanglerLayers`` template, which is a simple layer architecture 
+  of rotations and CNOT nearest-neighbour entanglers. 
+  [(#555)](https://github.com/XanaduAI/pennylane/pull/555)
+  
+  .. figure:: ../../_static/templates/layers/basic_entangler.png
+    :align: center
+    :width: 40%
+    :target: javascript:void(0);
```
Comment (josh146) [inline .github/CHANGELOG.md]: This is a markdown file, not ReST! So the images need to use markdown syntax _or_ html. Probably html is best, i.e.,
```html
  <img src="https://pennylane.readthedocs.io/en/latest/_images/qaoa_layers.png"
  width=70%></img>
```

Comment (mariaschuld) [reply .github/CHANGELOG.md]: Oh, of course. Strangely enough, it worked before...

Comment (mariaschuld) [reply .github/CHANGELOG.md]: The simplified_two_design figure link will be broken until the PR is merged into master...

File: tests/test_templates_layers.py (hunk 13 above)
Comment (antalszava) [reply tests/test_templates_layers.py]: Oh, you're right! Missed that part here :)

File: pennylane/init.py (hunk 7 above)
Comment (antalszava) [reply pennylane/init.py]: Oh I see!, I think you're right, that would be a good call

File: pennylane/templates/layers/simplified_two_design.py (hunk 6 above)
Comment (antalszava) [reply pennylane/templates/layers/simplified_two_design.py]: Looks good to me!

File: pennylane/templates/layers/simplified_two_design.py (hunk 4 above)
Comment (antalszava) [reply pennylane/templates/layers/simplified_two_design.py]: Right! I like the rewording, one thing that would be nice to include still is that these would be `qml.RY`s instead of `qml.PauliY`s :) (therefore added the word `rotation`, but I'm happy with any comparable rephrasing here).
